OSDOCS-8015: Adding a total resource usage table

mpandya15 opened the pull request:
[OSDOCS#8015]: Adding a total resource usage table

Version(s): 

Issue: https://issues.redhat.com/browse/OSDOCS-8015

Link to docs preview:
https://77127--ocpdocs-pr.netlify.app/openshift-enterprise/latest/observability/network_observability/configuring-operator.html

QE review:
- [x] QE has approved this change.
<!--- QE approval is required to merge a PR except for changes that do not impact the meaning of the docs. --->

Additional information:
Table mockup: https://docs.google.com/document/d/1xyBeIehZwyV_p7GkP9uJbr1ay5vFFcWsZwi3SENtB70/edit?usp=sharing

JoeAldinger commented:
/retest

jotak commented:
Hi @mpandya15 @skrthomas 
FYI here's more recommendations after running fine-tuning tests: https://issues.redhat.com/browse/NETOBSERV-1702?focusedId=24983810&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-24983810

> 
>    10-nodes cluster: set cacheMaxSize to 50000
>    20-nodes cluster: use 6 FLP replicas
>    65-nodes cluster: use 12 FLP replicas
>    120-nodes cluster: use 18 FLP replicas

cc @memodi

memodi commented:
> > 10-nodes cluster: set cacheMaxSize to 50000
> > 25-nodes cluster: use 6 FLP replicas
> > 65-nodes cluster: use 12 FLP replicas
> > 120-nodes cluster: use 18 FLP replicas

cc @stleerh 
thanks @jotak. Any reason why we'd recommend cacheMaxSize to 50K? I thought you mentioned no noticeable improvements were observed altering the cacheMaxSIze.

jotak commented:
@memodi it's extrapolated from my 6-nodes cluster, using my own tests (where fine-tuning leads me to 10K size). I don't think we have a test procedure for 10 nodes, do we?

memodi commented:
> I don't think we have a test procedure for 10 nodes, do we?

that's right, we don't. I am thinking several of our customers could fall into this category where they could have several smaller clusters < 10 nodes and based on these recommendation it would mean they'd have to change from default cacheMaxSize to 50k in flowcollector, is that something we want?

jotak commented:
@memodi they don't have to, this is only recommendations, and ideally they should do their own tuning. But they can definitely start with that. cf https://netobserv.io/documents/blogs/agent_metrics_perf/

mpandya15 commented:
> Hi @mpandya15 @skrthomas FYI here's more recommendations after running fine-tuning tests: https://issues.redhat.com/browse/NETOBSERV-1702?focusedId=24983810&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-24983810
> 
> > 10-nodes cluster: set cacheMaxSize to 50000
> > 25-nodes cluster: use 6 FLP replicas
> > 65-nodes cluster: use 12 FLP replicas
> > 120-nodes cluster: use 18 FLP replicas
> 
> cc @memodi

Hi Joel, thanks for bringing this up. We plan to include the `cacheMaxFlows` updates in a separate PR to update the existing info in the resource recommendations table.

skrthomas commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,42 @@
+
+:_mod-docs-content-type: REFERENCE
+[id="network-observability-total-resource-usage-table_{context}"]
+= Resource considerations
```
```suggestion
= Total average memory and CPU usage
```

skrthomas commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,42 @@
+:_mod-docs-content-type: REFERENCE
+[id="network-observability-total-resource-usage-table_{context}"]
+= Resource considerations
+The following table outlines examples of total resource usage for clusters with a sampling size of 1.
```
```suggestion
The following table outlines averages of total resource usage for clusters with sampling sizes of 1, 50, and 400.
```

skrthomas commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,42 @@
+The examples outlined in the table demonstrate scenarios that are tailored to specific workloads. Consider each example only as a baseline from which adjustments can be made to accommodate your workload needs.
+====
+
+.Total resource usage
```
```suggestion
.Total average resource usage
```

skrthomas commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,42 @@
+1. Tested with AWS M6i instances.
+2. In addition to this worker and its controller, 3 infra nodes (size `M6i.12xlarge`) and 1 workload node (size `M6i.8xlarge`) were tested.
+--
+Summary: Total resource usage of NetObserv (Agents+FLP+Kafka+Loki), for a 120 nodes cluster with an average traffic of 2.5 GBps, configured with sampling 1, is: 7.4 CPU cores (out of 1920), 39.82 GB. It generated an average of 17K flows per second.
```
```suggestion
Summary: Average total resource usage of Network Observability (Agents+FLP+Kafka+Loki), for a 120 nodes cluster with an average traffic of 2.5 GBps, configured with sampling 1, is: 7.4 CPU cores (out of 1920), 39.82 GB. It generated an average of 17K flows per second.
```

skrthomas reviewed (commented):
A few contextual clarifications. Looking really good!

memodi reviewed (changes requested):
thanks @mpandya15 , dropping couple of minor comments.

jotak commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,42 @@
+.Total average resource usage
+[options="header"]
+|===
+| Sample size | Parameters | Small (25 nodes)  | Medium (65 nodes) ^[2]^ | Large (120 nodes) ^[2]^
```
I think we need to use different headers for the columns. The main driver is not the size of the cluster, but it's the test being run. So it should be perhaps something like _Test 1_, _Test 2_, _Test 3_, with the cluster size in parenthesis. I'm insisting on this because otherwise, the numbers sometimes don't make sense. For instance, with sampling 400 we see that netobserv uses more CPU than on a medium cluster. That sounds wrong, but it's just that the test scenario is completely different.

jotak replied:
cc @memodi

mpandya15 replied:
Updated the headers to _Test 1, Test 2, Test 3_ along with the corresponding cluster sizes in parentheses.

memodi replied:
@jotak @mpandya15 - I am fine with having as Test 1, 2, 3 headers , I thinking that may be too abstract though, do we need to provide some visibility as to what each _Test_ is about?

skrthomas replied:
I agree @memodi . Its my feeling that Test 1, 2, and 3 and S, M and L are both generic naming conventions. But i don't think we can say node-density-heavy, ingress-perf, and cluster-density-v2. Would a customer have any idea what those are? I would guess no, but not sure if there's an alternative way to categorize these.

skrthomas replied:
I agree @memodi . Its my feeling that Test 1, 2, and 3 and S, M and L are both generic naming conventions. But i don't think we can say node-density-heavy, ingress-perf, and cluster-density-v2. Would a customer have any idea what those are? I would guess no, but not sure if there's an alternative way to categorize these.

skrthomas replied:
Whatever we decide here wrt naming conventions, we will need to apply it to the existing Resource consideration table updates as well.

jotak replied:
or we could add a footnote on each test, with a short description:

- Test 1 focuses on pods density per node, each of them generating a low traffic, on a small size cluster
- Test 2 focuses on ingress traffic throughput, on a medium size cluster
- Test 3 also focuses on pods density (like Test 1), running on larger cluster and generating more traffic

jotak replied:
cc @stleerh in case you want to chime in here

memodi replied:
short description as suggested [here](https://github.com/openshift/openshift-docs/pull/77127#discussion_r1657146887) sounds good to me. If we want to give specific metrics that represents scale (as OCP perf doc does) we collect `namespaceCount/podStatusCount/serviceCount` metrics for node/cluster-density workloads. For ingress-perf, we probably would need to rely on a metric that represents throughput of bytes (`nWorkloadBytesProcessedPerMinuteCadvisor`?).

mpandya15 replied:
Hi @jotak, thanks for your input, I've updated the footnotes. In response to @memodi's comment here about adding metrics in footnotes, these would be good details to add but more certainty would be needed around the `ingress-perf` metrics. Just needed to clarify whether this is something you recommend adding to the existing docs or a detail for future reference? Thank you! 😊

jotak commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,42 @@
+:_mod-docs-content-type: REFERENCE
+[id="network-observability-total-resource-usage-table_{context}"]
+= Total average memory and CPU usage
+The following table outlines averages of total resource usage for clusters with a sampling sizes of 1, 50, and 400.
```
Not sure if "sampling size" is correct , I would tend to say "sampling value" and not consider it as a _size_, wdyt? Same for the table header where it's "Sample size".

An alternative would be "Sampling ratio" and present it as 1:1, 1:50, 1:400

skrthomas replied:
I like sampling value instead of sampling size here. Thanks for the options for different wording :)

mpandya15 replied:
Thank you! Updated the headers and the paragraph with the "Sampling value" terminology.

jotak commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,42 @@
+1. Tested with AWS M6i instances.
+2. In addition to this worker and its controller, 3 infra nodes (size `M6i.12xlarge`) and 1 workload node (size `M6i.8xlarge`) were tested.
+--
+Summary: Average total resource usage of Network Observability (Agents+FLP+Kafka+Loki), for a 120 nodes cluster with an average traffic of 2.5 GBps, configured with sampling 1, is: 7.4 CPU cores (out of 1920), 39.82 GB. It generated an average of 17K flows per second.
```
This sentence doesn't correspond to the numbers presented above. We can either adapt it, or completely remove it?

mpandya15 replied:
@jotak @memodi Thanks for the clarification! Since the table is more comprehensive than just sampling value of 1, the summary now provides a generic statement that this table only presents the average total resource usage of Network Observability (Agents+FLP+Kafka+Loki). 

CC: @skrthomas :)

mpandya15 commented:
Hey @memodi @skrthomas @jotak @stleerh, I've addressed your feedback and made the required changes. PTAL at my latest updates and let me know if there's other changes. Thank you! 😊

memodi commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,44 @@
+|===
+
+--
+1. Test 1 focuses on pods density per node, each node generating low traffic on a small size cluster.
+2. Test 2 focuses on ingress traffic throughput on a medium size cluster.
+3. Test 3 focuses on pods density, generating more traffic on a large size cluster.
```
are these suppose to be numbered bullets or of citation style?

skrthomas replied:
@memodi The 1, 2, and 3 under the table are the footnotes for [1], [2], and [3] for each Test X column heading. Is that what you mean?

skrthomas replied:
Previously, we used to be able to use a `[.small]` attribute in asciidoc to make the footnotes look small, like actual footnote size. But (doc tools woes 😢 ) the Customer Portal docs don't render the `[.small]` attribute. Even though docs.openshift.com renders it fine, our contributing guide states not to use it because it isn't fully supported in all the places we publish. Not sure if you're wondering why the numbers look so huge, like code callouts.

memodi replied:
yes, that's what I was referring too, I thought it needs to be same of style [1], [2] or [3] as they were citied for e.g. `Test 1 (25 nodes) ^[1]^`

memodi commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,44 @@
+|===
+
+--
+1. Test 1 focuses on pods density per node, each node generating low traffic on a small size cluster.
```
```suggestion
1. Test 1 focuses on pods density per node, each pod generating low amounts of traffic on a small size cluster.
```

memodi commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,44 @@
+--
+1. Test 1 focuses on pods density per node, each node generating low traffic on a small size cluster.
+2. Test 2 focuses on ingress traffic throughput on a medium size cluster.
+3. Test 3 focuses on pods density, generating more traffic on a large size cluster.
```
```suggestion
3. Test 3 focuses on pods density, generating heavy traffic on a large size cluster.
```

jotak reviewed (approved):
LGTM, thanks @mpandya15 !

jotak commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,44 @@
+--
+1. Test 1 focuses on pods density per node, each pod generating low amounts of traffic on a small size cluster.
+2. Test 2 focuses on ingress traffic throughput on a medium size cluster.
+3. Test 3 focuses on pods density, generating heavy traffic on a large size cluster.
```
in fact we shouldn't say "heavy traffic" for Test 3, it's not heavy. I think at first my suggestion was to just say "more traffic than in Test 1"

skrthomas commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,44 @@
+3. Test 3 focuses on pods density, generating heavy traffic on a large size cluster.
+--
+
+Summary: This table shows average total resource usage of Network Observability (Agents+FLP+Kafka+Loki).
```
I think that "With" and "Without" are clearer here for accessibility considerations (thinking of screen readers).

skrthomas commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,44 @@
+:_mod-docs-content-type: REFERENCE
+[id="network-observability-total-resource-usage-table_{context}"]
+= Total average memory and CPU usage
+The following table outlines averages of total resource usage for clusters with a sampling value of 1, 50, and 400.
```
```suggestion
The following table outlines averages of total resource usage for clusters with a sampling value of 1, 50, and 400 for 3 different tests: `Test 1`, `Test 2`, and `Test 3`. The tests differ in the following ways:
- Test 1 and 3 take into account the `namespaceCount`, `podStatusCount` and `serviceCount` metrics , and represents use cases with a high number of workloads.
- Test 2 takes into account a high ingress traffic throughput.

Since different types of cluster use cases are exemplified in the different tests, these numbers cannot be linearly compared side-by-side but instead used as a benchmark to evaluate what your personal cluster usage may be.
```

skrthomas replied:
Based on [Slack](https://redhat-internal.slack.com/archives/C02939DP5L5/p1720812777670899?thread_ts=1720793630.658669&cid=C02939DP5L5) thread, we need to add more context about whats in the tests as a kind of "expectation-setting" to introduce the table.

skrthomas replied:
cc @stleerh @memodi wdyt about adding some expectation-setting to the introduction, like this? Anything you would add here?

memodi replied:
@skrthomas thanks, overall it looks good to me. Instead of listing each prometheus metric names that we defined in our tests, I'd say `... take into account total number of Namespace, Pods and Services in OCP cluster ... with a high number of workloads for a given cluster size`

skrthomas replied:
Ok, cool. That makes sense, thank you!

jotak replied:
That looks good to me, thanks @skrthomas 
I think given the tests that we run, this is well formulated. But still, this is IMO not what users would expect when they look for reference values. I think for 1.7 we should run consistent tests for different cluster sizes (like NDH-cni) for this doc - even if we continue to run other tests internally for perf regression tracking.

skrthomas replied:
@jotak  so if I understand, this is as good as we can give with the tests that we run today? And there's future potential to run consistent tests for different cluster sizes?

jotak replied:
Yes exactly, I hope in 1.7 we can simplify the doc by showing data with the same test used for each cluster size

skrthomas replied:
Do you think we should wait to publish this until 1.7, and wait on those updates? Or should we add these now?

skrthomas commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,44 @@
+| Total NetObserv RSS (Memory) Usage | 8.21 GB | 16.02 GB | 17.44 GB
+| *Without Loki* 3+|
+| Total NetObserv CPU Usage | 1.31 | 1.06 | 1.83
+| Total NetObserv RSS (Memory) Usage | 7.01 GB | 10.70 GB | 13.26 GB
```
Mehul said one decimal place is fine, so we can go to one.

skrthomas replied:
So when Manushri went to round these, there's a couple of times where this creates too similar looking data. For example, *Without Loki*
| Total NetObserv CPU Usage | 2.4 | 2.4 | 5.6

I think for this reason, we should stick to 2 decimal places. 

cc @mpandya15

jpinsonneau replied:
Maybe we could add `total % CPU usage` and `total % memory usage` in seperate rows or inside the same cells ? 

That was a question that often came discussing with customers.

jpinsonneau replied:
Also, should we mention prometheus impact here ? cc @jotak

skrthomas replied:
> Maybe we could add `total % CPU usage` and `total % memory usage` in seperate rows or inside the same cells ?
> 
> That was a question that often came discussing with customers.

@jpinsonneau I was chatting w/ @memodi about this, and he says we don't have these metrics now, but it will be something we can capture in future release.

jpinsonneau replied:
Can't we calculate the totals from the nodes hardware specs ? :thinking: 
Then it's a simple rule of three

memodi replied:
@jpinsonneau I don't know if we want total Netobserv % CPU/RSS Usage as a function of available on nodes OR as a function of cluster usage - from discussions with @stleerh we need latter (which I think makes sense) for which new prometheus queries needs to be added. Also, if we look w.r.t  nodes hardware it would be minuscule usage since that would be dependent on number of nodes in a cluster we obviously have several in each of our tests.

skrthomas replied:
For the Prometheus impact, if @jotak agrees, we can add a note like:
[IMPORTANT]
====
Metrics exported to Prometheus can impact the resource usage, so its important to be aware of the cardinality of the values you're exporting. For more information, see Flows format reference. 
====

Then link to the Flows format reference

jotak replied:
@skrthomas sgtm!

stleerh replied:
@memodi @jpinsonneau It should be with respect to cluster usage.  If it's based on availability, you can allocate a lot of resources without using it so that the percent used by NetObserv is low.

memodi commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,51 @@
+|===
+
+--
+1. Test 1 focuses on pods density per node, each pod generating low amounts of traffic on a small size cluster.
+2. Test 2 focuses on ingress traffic throughput on a medium size cluster.
+3. Test 3 focuses on pods density, generating more traffic than Test 1 on a large size cluster.
+--
```
@mpandya15 I think we can remove this now that you have provided detailed info about these tests before the table ?

memodi commented:
thanks @mpandya15

/label qe-approved

mpandya15 commented:
/label peer-review-needed

mletalie reviewed (commented):
/label peer-review-in-progress

mletalie commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,51 @@
+The following table outlines averages of total resource usage for clusters with a sampling value of 1, 50, and 400 for 3 different tests: `Test 1`, `Test 2`, and `Test 3`. The tests differ in the following ways:
+
+- `Test 1` takes into account the total number of namespace, pods and services in {product-title} cluster, places load on the eBPF agent, and represents use cases with a high number of workloads for a given cluster size. For example, `Test 1` consists of 76 Namespaces, 5153 Pods, and 2305 Services.
+- `Test 2` takes into account a high ingress traffic throughput.
```
Could "throughput" be replaced with "capacity" ? Not sure, but tbh had to look up definition of that word. :)

jotak replied:
then maybe "bandwidth" rather than "capacity", would that work for you?

mletalie replied:
Hello @jotak,
   Thanks for the response. I think either would work...and to be honest, "throughput" was not a term that I was familiar with...maybe others would be, just trying to make it as clear as possible. Thanks!

mpandya15 replied:
Switched it up with "volume." Thanks for the suggestion. :)

mletalie commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,51 @@
+
+The following table outlines averages of total resource usage for clusters with a sampling value of 1, 50, and 400 for 3 different tests: `Test 1`, `Test 2`, and `Test 3`. The tests differ in the following ways:
+
+- `Test 1` takes into account the total number of namespace, pods and services in {product-title} cluster, places load on the eBPF agent, and represents use cases with a high number of workloads for a given cluster size. For example, `Test 1` consists of 76 Namespaces, 5153 Pods, and 2305 Services.
```
```suggestion
- `Test 1` takes into account the total number of namespace, pods and services in an {product-title} cluster, places load on the eBPF agent, and represents use cases with a high number of workloads for a given cluster size. For example, `Test 1` consists of 76 Namespaces, 5153 Pods, and 2305 Services.
```

mletalie commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,51 @@
+
+- `Test 1` takes into account the total number of namespace, pods and services in {product-title} cluster, places load on the eBPF agent, and represents use cases with a high number of workloads for a given cluster size. For example, `Test 1` consists of 76 Namespaces, 5153 Pods, and 2305 Services.
+- `Test 2` takes into account a high ingress traffic throughput.
+- `Test 3` takes into account the total number of namespace, pods and services in {product-title} cluster, places load on the eBPF agent on a larger scale than `Test 1`, and represents use cases with a high number of workloads for a given cluster size. For example, `Test 3` consists of 553 Namespaces, 6998 Pods, and 2508 Services.
```
```suggestion
- `Test 3` takes into account the total number of namespace, pods and services in an {product-title} cluster, places load on the eBPF agent on a larger scale than `Test 1`, and represents use cases with a high number of workloads for a given cluster size. For example, `Test 3` consists of 553 Namespaces, 6998 Pods, and 2508 Services.
```

mletalie commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,51 @@
+
+[NOTE]
+====
+Metrics exported to Prometheus can impact the resource usage. Cardinality values for the metrics can help determine how much resources are impacted.. For more information, see Flows format reference.
```
```suggestion
Metrics exported to Prometheus can impact the resource usage. Cardinality values for the metrics can help determine how much resources are impacted. For more information, see Flows format reference.
```

mletalie replied:
```suggestion
Metrics exported to Prometheus can impact the resource usage. Cardinality values for the metrics can help determine how much resources are impacted.. For more information, see "Network Flows format" in the Additional resources section.
```

mletalie commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,51 @@
+The examples outlined in the table demonstrate scenarios that are tailored to specific workloads. Consider each example only as a baseline from which adjustments can be made to accommodate your workload needs.
+====
+
+.Total average resource usage
```
Maybe reference each test in the table? In other words, instead of just providing sampling value as indicator of which test is which, instead identify test 1, 2 and 3?

mletalie reviewed (commented):
/label peer-review-done
/remove-label peer-review-in-progress
/remove-label peer-review-needed
/lgtm

mletalie commented:
Hello @mpandya15,
    Just a couple of comments/suggestions, otherwise LGTM. Good job! Thanks.

mpandya15 commented:
/label merge-review-needed

skrthomas commented on modules/network-observability-total-resource-usage.adoc:
```diff
@@ -0,0 +1,51 @@
+
+Since different types of cluster use cases are exemplified in the different tests, the numbers in this table cannot be linearly compared side-by-side, but instead are intended to be used as a benchmark for evaluating your personal cluster usage.
+
+[NOTE]
+====
+Metrics exported to Prometheus can impact the resource usage. Cardinality values for the metrics can help determine how much resources are impacted. For more information, see "Network Flows format" in the Additional resources section.
+====
```
Since we have these two admonitions back-to-back, which I know I am the one who said we should put these things in admonitions, maybe we can work one of them into the main text. Like I think the IMPORTANT can maybe be incorporated into the main text, then leave the NOTE. So that last paragraph would be like this:
> Since different types of cluster use cases are exemplified in the different tests, the numbers in this table cannot be linearly compared side-by-side, but instead are intended to be used as a benchmark for evaluating your personal cluster usage. The examples outlined in the following table demonstrate scenarios that are tailored to specific workloads. Consider each example only as a baseline from which adjustments can be made to accommodate your workload needs. 

> [NOTE]
====
Metrics exported to Prometheus.....
====

skrthomas replied:
Here's the SSG reference for not stacking admonitions: https://redhat-documentation.github.io/supplementary-style-guide/#admonitions

skrthomas reviewed (commented):
One minor thing to tweak, then I think we're good to go! Nice work!

skrthomas commented:
/cherrypick enterprise-4.12

skrthomas commented:
/cherrypick enterprise-4.13

skrthomas commented:
/cherrypick enterprise-4.14

skrthomas commented:
/cherrypick enterprise-4.15

skrthomas commented:
/cherrypick enterprise-4.16
